### DISCUSSION TITLE
Can't connect to jsbsim server

### DISCUSSION THREAD

**Discussion by achodoglugil** [JSBSim-Team/jsbsim#693] [Q&A]
I set the input parameters like this in c1723.xm
![input](https://user-images.githubusercontent.com/74592595/182802102-e75f7864-0e2b-4f87-b3ce-059358bf8066.PNG)
l
i run the command .\JSBSim --script=scripts/c1723.xml --realtime --suspend and then from another terminal i run the command "telnet localhost 1137" and doesnt get the message "connected to jsbsim server". I just get a blank terminal and cant type anything

**Comment by seanmcleod** [comment]
As a quick test what happens if you don't include `--suspend` when you run JSBSim?

**Reply by achodoglugil** [reply]
nothing changes

**Comment by seanmcleod** [comment]
Ah, I've just looked at your screenshot again. You've placed the `<input>` element in the wrong file, you've placed it in the script file whereas it needs to be placed in the aircraft configuration file.

> JSBSim has the capability to output data at runtime via a socket. The capability also exists to connect to JSBSim via socket for data input. The capability is more suited for interaction and debugging or investigations more than for interprocess communications. The **aircraft configuration file** may contain the optional element that defines how an outside process can connect to a running instance of JSBSim:

**Reply by achodoglugil** [reply]
what is the aircraft configuration file for c1723.xml can you help me with this

**Comment by seanmcleod** [comment]
It's listed 1 line below in your screenshot, i.e.

`<use aircraft="c172x".... />`

**Reply by achodoglugil** [reply]
oh thank you very much. I put the input element into the c172x.xml file like this and deleted it from the c1723.xml script
![Ekran Alıntısı](https://user-images.githubusercontent.com/74592595/182811166-8cf13e27-79aa-47ea-89ff-ed2efae62583.PNG)
still having the same problem, when i run the jsbsim with the "jsbsim --script=scripts/c1723.xml --realtime --suspend" command from another terminal i use "telnet localhost 1137" command, it promts a message "connecting localhost..." and blank terminal appears. I also tried without --suspend and nothing changed.

**Reply by bcoconni** [reply]
Works for me when the `--realtime` option is removed. It seems that the `--realtime` option is not compatible with the socket interface, most likely because it makes JSBSim spend most of its time calling `sleep()` which prevents it from responding to the socket (and `telnet`).

The steps I followed:
1. Uncommented the lines 1193-1195 in `aircraft/c172x/c172x.xml`
https://github.com/JSBSim-Team/jsbsim/blob/3f939fa7dc9f9d0671124d8deffc0d128019aa41/aircraft/c172x/c172x.xml#L1193-L1195
1. Run JSBSim with the `--suspend` option
```bash
> ./JSBSim scripts/c1723.xml --suspend
```
3. Launch `telnet` in another terminal:
```bash
> telnet localhost 1137
```

**Reply by achodoglugil** [reply]
i followed all the steps you said and still getting a blank terminal. I don't know what i am doing wrong everything seems fine but can't establish the connection, even checked the firewall permissions and they seem fine too.

**Reply by bcoconni** [reply]
What do you mean by "getting a blank terminal" ? Because, on my side, when the connection fails, I get the following message:
```bash
> telnet localhost 1137
Trying 127.0.0.1...
telnet: Unable to connect to remote host: Connection refused
```

**Reply by achodoglugil** [reply]
let me show you with screenshots, when i enter the telnet localhost 1137 command i get a message like this
![terminal](https://user-images.githubusercontent.com/74592595/183067769-d1b22841-508a-4d34-834f-215a6dfd38f2.png)
and then the terminal becomes like this and cant type or do anything
![terminal2](https://user-images.githubusercontent.com/74592595/183067852-dc165c89-9925-4951-9aa6-fffc72364424.png)
and this is the screenshot of the report
![tcp](https://user-images.githubusercontent.com/74592595/183068031-0e37ced6-310c-482d-ac3a-8cd4a54e0786.PNG)

**Reply by achodoglugil** [reply]
bağlanıyor means connecting, sorry about that i use windows in turkish

**Comment by bcoconni** [comment]
Also, could you check that JSBSim reports the following output in the terminal where you ran the command `./JSBSim scripts/c1723.xml --suspend` ?
```text
Creating input TCP socket on port 1137
Successfully bound to TCP input socket on port 1137
```

**Comment by seanmcleod** [comment]
Hmm, so I can reproduce the issue with the same symptoms on Windows 11. Telnet connects but there is no output shown.

Used `netsat -a` to confirm the connection.

```sh
(base) C:\Users\Sean>netstat -a

Active Connections

  Proto  Local Address          Foreign Address        State
  TCP    0.0.0.0:135            ThinkCentre:0          LISTENING
  TCP    0.0.0.0:445            ThinkCentre:0          LISTENING
  TCP    0.0.0.0:1137           ThinkCentre:0          LISTENING
  TCP    0.0.0.0:2179           ThinkCentre:0          LISTENING
  TCP    0.0.0.0:2869           ThinkCentre:0          LISTENING
  TCP    0.0.0.0:3389           ThinkCentre:0          LISTENING
  TCP    0.0.0.0:5040           ThinkCentre:0          LISTENING
  TCP    0.0.0.0:49664          ThinkCentre:0          LISTENING
  TCP    0.0.0.0:49665          ThinkCentre:0          LISTENING
  TCP    0.0.0.0:49666          ThinkCentre:0          LISTENING
  TCP    0.0.0.0:49667          ThinkCentre:0          LISTENING
  TCP    0.0.0.0:49669          ThinkCentre:0          LISTENING
  TCP    0.0.0.0:49671          ThinkCentre:0          LISTENING
  TCP    0.0.0.0:49683          ThinkCentre:0          LISTENING
  TCP    127.0.0.1:1137         ThinkCentre:58408      ESTABLISHED
``` 

I'll run JSBSim in the debugger to try and see what's happening. Is it sending the text `connected etc...` and if so is there some terminal issue which isn't displaying the text sent?

**Comment by seanmcleod** [comment]
Hmm, so the first thing I noticed is that in order to get to `send("Connected to JSBSim server..")` that means `sckt_in` needs to be <= 0. However `sckt_in` as seen in the Locals list is an unsigned 64bit int.

![image](https://user-images.githubusercontent.com/824856/183144376-9d24dbe1-2cbb-4f52-923c-2029ac1fa976.png)


From `FGfdmSocket.h`

```c++
#if defined(_MSC_VER) || defined(__MINGW32__)
  SOCKET sckt;
  SOCKET sckt_in;
#else
  int sckt;
  int sckt_in;
#endif
```

From `WinSock2.h`

```c++
/*
 * The new type to be used in all
 * instances which refer to sockets.
 */
typedef UINT_PTR        SOCKET;
```

And as mentioned here - https://docs.microsoft.com/en-us/windows/win32/winsock/socket-data-type-2 

> This means, for example, that checking for errors when the [socket](https://docs.microsoft.com/en-us/windows/desktop/api/Winsock2/nf-winsock2-socket) and [accept](https://docs.microsoft.com/en-us/windows/desktop/api/Winsock2/nf-winsock2-accept) functions return should not be done by comparing the return value with –1, or seeing if the value is negative (both common and legal approaches in UNIX). Instead, an application should use the manifest constant INVALID_SOCKET as defined in the Winsock2.h header file. For example:

Which makes me wonder how this code ever worked on Windows?

Taking a quick glance at the git history for `FGfdmSocket.h` it looks like the type for `sckt` and `sckt_in` has been changed/toggled between `int` and `SOCKET` a couple of times.

The most recent change was on 1/18/2020 which changed them back to `SOCKET` for Windows - https://github.com/JSBSim-Team/jsbsim/commit/41bc6592c1dae1c33dc398a3c7e7e91a4b54e2c0

**Reply by bcoconni** [reply]
Good catch @seanmcleod :+1:

> Which makes me wonder how this code ever worked on Windows?

Well, it seems it never did since the commit 41bc6592c1dae1c33dc398a3c7e7e91a4b54e2c0 has been pushed. Since the test `TestInputSocket` is broken on Windows the regression went unnoticed.

> Taking a quick glance at the git history for `FGfdmSocket.h` it looks like the type for `sckt` and `sckt_in` has been changed/toggled between `int` and `SOCKET` a couple of times.

The problem is that the [function `accept()` returns a `SOCKET` under the Windows](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-accept) while [it returns an `int` on BSD/Unix](https://man7.org/linux/man-pages/man2/accept4.2.html)

> The most recent change was on 1/18/2020 which changed them back to `SOCKET` for Windows - https://github.com/JSBSim-Team/jsbsim/commit/41bc6592c1dae1c33dc398a3c7e7e91a4b54e2c0

Oops, I did push this commit. Apparently, [the change was motivated by the removal of warnings issued by MSVC](https://github.com/JSBSim-Team/jsbsim/pull/249/commits/cd619e0854998c46abcbe7f87a4691ab784bedf9). However as you pointed out, on Windows the failure of `accept()` is not reported by a negative value but by the special value `INVALID_SOCKET`. So on Windows we should check that `sckt_in` is not equal to `INVALID_SOCKET` rather than checking it is positive. Actually there are a lot of places in `FGfdmSocket.cpp` where the wrong comparison is made on `sckt_in` and `sckt`.

**Comment by seanmcleod** [comment]
As a quick test I changed the type from `SOCKET` to `int` and recompiled, bunch of warnings but it compiled.

```sh
1>FGfdmSocket.cpp
1>C:\source\jsbsim\src\input_output\FGfdmSocket.cpp(124,16): warning C4244: '=': conversion from 'SOCKET' to 'int', possible loss of data
1>C:\source\jsbsim\src\input_output\FGfdmSocket.cpp(168,18): warning C4244: '=': conversion from 'SOCKET' to 'int', possible loss of data
1>C:\source\jsbsim\src\input_output\FGfdmSocket.cpp(178,18): warning C4244: '=': conversion from 'SOCKET' to 'int', possible loss of data
1>C:\source\jsbsim\src\input_output\FGfdmSocket.cpp(204,27): warning C4244: '=': conversion from 'SOCKET' to 'int', possible loss of data
1>C:\source\jsbsim\src\input_output\FGfdmSocket.cpp(246,23): warning C4244: '=': conversion from 'SOCKET' to 'int', possible loss of data
1>FGInput.cpp
```

And sure enough now I can telnet to JSBSim successfully.

**Reply by bcoconni** [reply]
Unfortunately, changing the type of `sckt` and `sckt_in` to `int` did not make any difference on my side: JSBSim does not respond to telnet in both cases. It only works on Linux.

**Reply by bcoconni** [reply]
> Unfortunately, changing the type of `sckt` and `sckt_in` to `int` did not make any difference on my side: JSBSim does not respond to telnet in both cases. It only works on Linux.

I finally managed to get the socket connection working when setting `sckt` and `sckt_in` as `int`. I'm currently preparing a fix to keep `sckt` and `sckt_in` being of the `SOCKET` type and having a working socket connection.

**Reply by achodoglugil** [reply]
Thanks for the replies, i will try the fixes as soon as possible and let you know if it worked for me.

**Reply by achodoglugil** [reply]
I changed the `sckt` and `sckt_in` as `int` but still cant get the socket connection working. I made the changes on `FGfdmSockect.h` , do i need to change anything else

**Reply by seanmcleod** [reply]
That's all I changed to get it to work.

**Reply by achodoglugil** [reply]
After rebuilding i get to work it, thanks for all the help.

**Comment by bcoconni** [comment]
FYI, I have created issue #698 upon the problems reported in this discussion.
